src/libgap-api.c

```diff
@@ -0,0 +1,136 @@
+// LibGAP API - API for using GAP as shared library.
+
+#include "libgap-api.h"
+
+#include "bool.h"
+#include "calls.h"
+#include "gapstate.h"
+#include "gvars.h"
+#include "intobj.h"
+#include "opers.h"
+#include "plist.h"
+#include "stringobj.h"
+
+Obj FuncREAD_ALL_COMMANDS(
+    Obj self, Obj instream, Obj echo, Obj capture, Obj resultCallback);
+
+// Setup and initialisation
+void GAP_Initialize(int          argc,
+                    char **      argv,
+                    char **      env,
+                    CallbackFunc markBagsCallback,
+                    CallbackFunc errorCallback)
+{
+    InitializeGap(&argc, argv, env);
+    SetExtraMarkFuncBags(markBagsCallback);
+    STATE(JumpToCatchCallback) = errorCallback;
+}
+
+// TODO: This function is far from completely finalizing a GAP session
+//       it potentially has to close all open file descriptors, kill
+//       subprocesses, etc
+void GAP_Finalize(void)
+{
+    FinishBags();
+}
+
+
+/*************************************************************************/
+/*** Integer *************************************************************/
+/*************************************************************************/
+
+Obj GAP_IntObj_Int(Int value)
+{
+    return INTOBJ_INT(value);
+}
+
+Int GAP_Int_IntObj(Obj value)
+{
+    return INT_INTOBJ(value);
+}
+
+/*************************************************************************/
+/*** Strings *************************************************************/
+/*************************************************************************/
+
+char * GAP_CSTR_STRING(Obj str)
+{
+    return CSTR_STRING(str);
+}
+
+Obj GAP_MakeString(char * str)
+{
+    return MakeString(str);
+}
+
+/*************************************************************************/
+/*** List ****************************************************************/
+/*************************************************************************/
+
+Obj GAP_NewPList(UInt length)
+{
+    return NEW_PLIST(T_PLIST, length);
+}
+
+void GAP_SetLenPList(Obj list, Int len)
+{
+    SET_LEN_PLIST(list, len);
+}
+
+void GAP_AssPList(Obj list, Int pos, Obj elem)
+{
+    AssPlist(list, pos, elem);
+}
+
+void GAP_SetElmPList(Obj list, Int pos, Obj elem)
+{
+    SET_ELM_PLIST(list, pos, elem);
```

Comment:
I wrote the following before, but due to updates to this PR, it is now hidden; but I still would like a reply by @sebasguts on this:

Honest open question: I wonder whether there is an actual reason (e.g. extreme performance needs?) for why this "API" exposes potentially dangerous low-level functions like `SET_ELM_PLIST`, or even `AssPlist` -- wouldn't it be more useful and safer to expose `ASS_LIST`, `ELM0_LIST`, `LEN_LIST`, etc.?

I just had a look at the sage code, and also at https://hackmd.io/emNi76svSWCh1fBeLKqPdA# (which @nthiery and me created together). Some notes:
* `LEN_PLIST` is used once in SageMath -- incorrectly, though, as it is also called on ranges, blists, strings -> should really use `LEN_LIST` instead
* as far as I can tell, no other GAP function with the string `plist` (upper or lower case) is used by SageMath, although a few more are imported (but then not used)

Perhaps @sebasguts "needs" some of them? Is there perhaps a code repository somewhere to look at?

Reply:
Also, @markuspf replied:
> Finding what the correct API is is, in my opinon, the name of the game for the immediate future of "LibGAP".

I have a clear opinion on this: such an API should overall aim to provide high-level APIs at first; and only add lower-level APIs if there is a strong justification that make the higher level APIs unsuitable (and even then, I'd first investigate making the higher level APIs better). In this case, I think it's a mistake to export all those plist accessors, when we already have generic list accessors. The only reason to use the plist ones is raw performance, but a sizeable portion of that is sacrificed by these non-inlined wrappers anyway. There may be other reasons to export plist-only APIs, but these then should be conscious decisions (and ideally explained in comments).

Reply:
That said, if you guys really, really want to expose these plist APIs (even though e.g. `GAP_SetElmPList` is impossible to use correctly via the provided API GAP_SetElmPList, since `CHANGED_BAG` is not exposed), fine, go ahead -- but I think it runs totally against the idea of creating a long-term stable API, as naturally lower level interfaces are much more likely to change than higher level ones.

Reply:
One more note regarding the symbols used by SageMath: a lot of the low level ones come from the function `gap_eval`. My suggestion here would be to try and replace most of it by using `READ_ALL_COMMANDS`, or a similar new function. To get somewhere quickly, I'd not worry about making the ideal generic function we all want in the long run, but just something that does precisely what SageMath needs; we then may have to touch this again later down the road, but I think that'll be much easier once we on the GAP side have clear insights on what you need. Perhaps @nthiery has thoughts on this, too?

Reply:
This is precisely what `GAP_EvalString` does (or, well, is supposed to do!)

Reply:
@fingolfin said

> Also, @markuspf replied:
> > Finding what the correct API is is, in my opinon, the name of the game for the immediate future of "LibGAP".

> I have a clear opinion on this: such an API should overall aim to provide high-level APIs at first;  and only add lower-level APIs if there is a strong justification that make the higher level APIs unsuitable (and even then, I'd first investigate making the higher level APIs better). In this case, I think it's a mistake to export all those plist accessors, when we already have generic list accessors. The only reason to use the plist ones is raw performance, but a sizeable portion of that is sacrificed by these non-inlined wrappers anyway. There may be other reasons to export plist-only APIs, but these then should be conscious decisions (and ideally explained in comments).

I completely agree. The API that is in `libgap-api.c` right now is what I stole from @sebasguts' code, the only "high level" function I added is `GAP_EvalString`.

I will update this PR with a basic API which I think is useful (informed by the hackmd document you hinted at, and an effort of mine of tying GAP into [racket](https://racket-lang.org) and [idris](https://idris-lang.org). Needless to say that feedback from serious users of the API (i.e. SageMath and Oscar) would be incredibly useful.

Reply:
Sorry for being late to the show. I would be perfectly happy with the lowercase-functions, actually. I am not very picky here, we would need to do the input checks done by the lowercase versions for OSCAR anyway. I doubt there is a speed difference when doing it from julia or from GAP. Code to look at can be found here: https://github.com/oscar-system/LibGAP.jl/blob/master/src/libgap.jl

Furthermore, I also think the way @markuspf did the new version is also good!